Hh new reactive

hhaensel opened the pull request:
New Reactive API which is backwards compatible but introduces public, private, readonly and jsfunctions modes for model fields.

 I create this PR to discuss the potential changes.

Current state:
- Variables with names ending with `_private` are excluded from the frontend.
- `Private{T}` is introduced as wrapper type for fields that don't go into the frontend, but I'm not sure whether this is the optimal solution
- `Reactive{T}` becomes a mutable struct with a mode field. `:public` is default and is identical to the current implementation, `:private` is excluded from the frontend. `:readonly` does not receive a watcher in the frontend. (Currently it can manually be changed from the frontend, but we should prevent that in the future). `:jsfunction` indicates that the variable contains either a JSONText with a js function or a `Dict` that possibly contains a js function. This variables are (currently) readonly.

Some background to js functions: 
- Why should we have it?
Sometimes it is very convenient to execute js functions in the frontend, e.g. for messenging, debugging or complex animations. Secondly, Apexcharts uses functions for axis formatting. (It is at least the only way I found to change the number of decimals in axis labels.)
- What is special about functions?
If js functions are initialised in the data section, they need to be defined with some `JSONText`, which is straight forward. But the dynamic change of a js function is tricky, as JSON does not provide a standard format. I have defined my own format after various inspiring readings and use the `revive` option in the json parser to construct functions in the frontend. This is invisible to the user and happens either in the `push!(model, ...)`routine or in the reactive listener that is attached during model setup.

Tasks to be done:
- `update!` needs to differentiate whether a property is private or not (see `setindex!` ...)
- readonly properties can currently be changed manually
- a switch enableing/disabling js function reviver
- support a selection of keys, e.g. a dictionary of regex patterns, for which the reviver is enabled

Proposal for other changes:
- use `setindex!` to change array parameters with notification and define a special method to replace the current setindex method
- modify the listener to better handle typed arrays

hhaensel commented:
I fixed the critical parts and changed the API for readonly and private.
Now, non-reactive fieldnames ending with "_" are readonly, fieldnames ending with "__" are private. That looks much nicer. I did this, because I couldn't find a wrapper type that was flexible enough. I still have the `Private{T}` type defined. @essenciary If you know a better way to implement it, please tell me. Otherwise I would propose to remove `Private` before role-out.

Moreover I fixed a bug in `update!()` that prevents updates from the frontend for non-reactive types. We need to fix that in master, if we don't want to role out the new API.
But I think, it's ready.

essenciary commented:
Thank you! 

The field naming convention approach has a few notable downsides - listing them together with possible ways to mitigate the issues: 

1. if people are not aware of this and use this naming for their fields, they will end up with rage inducing bugs. This can be addressed with a one time warning that `_` and `__` are used for read-only and private properties if such fields are detected. 
2.  it can clash with other naming constraints (people might need these endings for other things, like to match the name of a DB column). This can be addressed by having `_` and `__` in config. On a separate note, as a general best practice, we should not hard code such "magic strings" (https://softwareengineering.stackexchange.com/questions/365339/what-is-wrong-with-magic-strings).
3. from a stylistic perspective, if one needs to work with these fields a lot, it's unpleasant to be forced to use the `__` ending in all their code (people can set an alias for the field). 
4. do we need private fields at all? Do they need to be in the `ReactiveModel` if they are private? 

That being said, I'm of the opinion that a good working solution now is better than a perfect solution in one year. So I'm happy to merge it, but ideally we should put the magic strings (`_` and `__`) into vars so they can be read and written.

We can remove the `Private` API if it doesn't work - otherwise it will just confuse users looking into the code.

hhaensel commented:
Thanks fir the comments. That sounds all reasonable and feasable. I'll let you know when I'm finished

hhaensel commented:
@essenciary

1. I continued thinking about the API and I think we could make the `Private{T}` type working, if we agreed that these variables always have to be used like the reactive types, i.e. by using `[]`.
Alternatively we could just enhance the reactive type by another property `autoupdate` which decides whether a variable get's a listener during setup. Then we could easily offer two APIs, the reactive one and the basic one with pattern matching.

1. Why a private field? It is certainly convenient to store some properties with the model. My first candidate would be the channel which is stored with the model. That would also make the messenging much cleaner if we wouldn't have to store channel properties in the Genie.Settings if we think about including two models on a page. It would also allow us to define the push method without a channel, just `push!(model, field => value)`

hhaensel commented:
I revised the API, but I'm wondering, whether we should encode the mode properties similar as the `no_watcher` properties and not with a symbol. What do you think?
Anyhow, please test and comment.

essenciary commented:
@hhaensel Thank you so much! I'll run the tests/demos and see if I can provide any relevant input :) The integration is getting quite complex by now so you're the best to judge :)

hhaensel commented:
I don't ask for support in implementing, just for your advice in style 😉

hhaensel commented:
> Proposal for other changes:
> 
> * [ ]  use `setindex!` to change array parameters with notification and define a special method to replace the current setindex method
> * [ ]  modify the listener to better handle typed arrays

Do you have any opinion about this topic? I would rename the current  `setindex!(r::Reactive{T})` to `set_withoutlistener!()` and define
```
function setindex!(r::Reactive{T}, v::Any, index1, indices...)`
  r[][index, indices...] = v
  notify(r)
end
```
That way it would be quite easy to change single array elements. For the change of more than one element one would have to call notify manually.

hhaensel commented:
I added support for OffsetArrays. This is very handy, if you use bind array values to your elements, because you can chose 0-based indexing and have the same indexing in both front-end and backend.
Furthermore, I convert a single value to an array, if the field is an array type. This is convenient for some js elements that can either hold amn atomic value or an array, e.g. multi-sliders.

@essenciary Did you get on with your testing?

hhaensel commented:
There were some errors in the Reactive type setup so that it was not possible to use parametric types in the model. I got some inspiration by `Obervables.jl`

hhaensel commented:
BTW, I also added [support of typed arrays](https://github.com/GenieFramework/Stipple.jl/pull/43/commits/b6c6b853048cc2d8375f996a608de7b2b65580df#diff-39019d8ae899bcdd3758db28f205d5c546be80ab59858397e829ce82fa2e9f38R301-R302).

essenciary commented:
@hhaensel Thank you, looks great! 

The only think I'm not sure about is if it is worth having OffsetArrays as a dependency of core Stipple. Is it that widely used? (I never used it). Would it require that the users are aware of the existence of OffsetArrays and use the OffsetArrays API or is it transparent to the users? Can't it be a separate `StippleOffsetArrays` package that specializes some `convert` function?

essenciary commented:
@hhaensel I've merged a series of changes into master 
https://github.com/GenieFramework/Stipple.jl/commit/cd73554ba56387fba7feefd0f278d746aeadb9f1

Mostly doc comments thus completing the comments for Stipple core. 
Currently working on UI tests and building API docs in Stipple core. 

The only potentially breaking change is that I moved `@click` into StippleUI since it's Quasar functionality which is only in StippleUI. However, I set `@click` to be exported in StippleUI so assuming the code that used `@click` also used `StippleUI` it should be non-breaking.

hhaensel commented:
> The only thing I'm not sure about is if it is worth having OffsetArrays as a dependency of core Stipple. Is it that widely used? (I never used it). Would it require that the users are aware of the existence of OffsetArrays and use the OffsetArrays API or is it transparent to the users? Can't it be a separate `StippleOffsetArrays` package that specializes some `convert` function?

The user doesn't need to be aware of the package. Users that want to use OffsetArrays, can do it. If you think that we should not include the package, we can include parts of the package only, if the user loads OffsetArrays.

essenciary commented:
We should be very strict about what dependencies we add in Stipple core as these are used by all the Stipple libraries potentially leading to longer precompilation times, versioning issues, etc. Stipple core provides the low-level communication architecture between Julia and JS/Vue. 

I think we should move this kind of logic in specialized packages or user code. We could define a `convertfield(oldvalue, newvalue)` function which is invoked from `update!` and can be specialized to support any kind of types.

hhaensel commented:
Cool approach, I will manage that. I thought about doing this with update! itself. I might find a solution this weekend

hhaensel commented:
@essenciary Coming back to an old issue: could you give me an Example where you need the `parse()` function?

essenciary commented:
We use `parse` to convert generic data types into the specific ones used by the various Stipple types (each package which defines the types adds `parse` functions). There's quite a few: 

![image](https://user-images.githubusercontent.com/11292378/114281503-fee4f600-9a3e-11eb-94db-6c69e76b8db9.png)

hhaensel commented:
I reconfigured the value parsing and introduced `convertvalue()` in contrast to `Base.convert()`, because I need to read the value of the target variable and not the type only. Currently I have not yet separated OffsetArrays for easier testing. If you think the setup is ok, I will continue.

What do you think about the idea to use the package [`Requires.jl`](https://github.com/JuliaPackaging/Requires.jl) to support packages that are not part of the dependencies? In that case we would not need to file another package for 10 lines of code.

One last question. Why do you use `Base.parse()` for the type conversion. Wouldn't it be better to introduce a Stipple-owned parsing routine, e.g. `Stipple.parsetype()` in order to avoid type pirating?

essenciary commented:
Cool progress, thank you for the updates! 

Yes, let's give `Requires` a try and see how it goes. If good, we can extend it's usage for other cases. If not, we can refactor later. 

It wouldn't be type piracy as `Base.parse` in this case is designed to be specialized for new types provided by the Stipple packages. And this is the exact use case for `Base.parse`. If a `Base.parse` method is already defined for that type it will just be used, no need to pirate/specialize anymore. It's about tapping into existing APIs/conventions.

essenciary commented:
@hhaensel I think we should try to merge this, it's becoming very complex as it covers too many different topics. We can move some of the threads into separate issues so we can work on more granular changes. This PR will introduce so many changes that if anything breaks, it will be hard to trace the issues and fix them.

hhaensel commented:
I agree, I'll try to get it done tonight!

essenciary commented:
> 
> 
> I agree, I'll try to get it done tonight!

Thank you. No rush, but best not to add new features here please.

hhaensel commented:
@essenciary looks as if we are set for larger testing.

Yesterday night I added support of indexing into Reactive fields with notification and I added `Requires.jl` and changed OffsetArray support accordingly.

A very rough testing of the API can now be found in the [StippleDemos](https://github.com/GenieFramework/StippleDemos). `New API3 Demo Draft.jl`

Would be great to have a feedback!

essenciary commented:
@hhaensel are you on any messaging platform (Gitter [Genie's], Slack [Julia's], Discord [Gennie's], WhatsApp, etc) so we can maybe follow up more real time on quick things?

essenciary commented:
@hhaensel also, what branch should I test?

essenciary commented:
My plan is to test using the current #master of Stipple, StippleUI, and StippleCharts and tag new versions - so we have a final version for the current API. 

Then test with the new API and once all is good we tag a new version. 

Anything wrong with this approach?

hhaensel commented:
Ok for me. Feel free to merge!

hhaensel commented:
> @hhaensel also, what branch should I test?

Well, the one that this PR is about, so `hh-newReactive`
